lib/resty/core/regex.lua

```diff
@@ -479,6 +480,76 @@ function ngx.re.find(subj, regex, opts, ctx, nth)
 end
 
 
+local function iterate_re_gmatch(self)
+    local compiled = self._compiled
+    local subj = self._subj
+    local subj_len = self._subj_len
+    local flags = self._flags
+    local pos = self._pos
+
+    if not pos then
+        -- The iterator is exhausted.
+        return nil
+    end
+
+    local rc = C.ngx_http_lua_ffi_exec_regex(compiled, flags, subj,
+                                             subj_len, pos)
+
+    if rc == PCRE_ERROR_NOMATCH then
+        if not self._compile_once then
+            destroy_compiled_regex(compiled)
+        end
+        self._pos = nil
+        return nil
+    end
+
+    if rc < 0 then
+        if not self._compile_once then
+            destroy_compiled_regex(compiled)
+        end
+        self._pos = nil
+        return nil, "pcre_exec() failed: " .. rc
+    end
+
+    if rc == 0 then
+        if band(flags, FLAG_DFA) == 0 then
+            self._pos = nil
+            return nil, "capture size too small"
+        end
+
+        rc = 1
+    end
+
+    self._pos = compiled.captures[1]
+    return collect_captures(compiled, rc, subj, flags)
+end
+
+
+do
+    local re_gmatch_iterator_mt = { __call = iterate_re_gmatch }
+
+    function ngx.re.gmatch(subj, regex, opts)
+        subj  = tostring(subj)
+
+        local compiled, compile_once, flags = re_match_compile(regex, opts)
+        if compiled == nil then
+            -- compiled_once holds the error string
+            return nil, compile_once
+        end
+
+        local re_gmatch_iterator = new_tab(0, 6)
```

Comment:
Hm, maybe not, if we use several instances of this iterator on different subjects/regexps at once, as it seems to be a supported use-case... Too bad.